File: mobile/lib/src/features/auth/data/service/auth_service.dart
import 'package:seattle_pulse_mobile/src/core/api/api_client.dart';
import 'package:seattle_pulse_mobile/src/core/api/api_endpoints.dart';
import 'package:seattle_pulse_mobile/src/features/auth/data/service/user_secure_storage.dart';
import '../models/user_model.dart';
import '../models/register_response_model.dart';
import '../models/verify_account_response_model.dart';

class AuthService {
  final ApiClient api;

  AuthService(this.api);

  Future<RegisterResponseModel> registerUser({
    required String firstName,
    required String lastName,
    required String username,
    required String email,
    required String password,
    required bool acceptedTerms,
  }) async {
    final response =
        await api.post('${ApiEndpoints.register}?isMobile=true', data: {
      "first_name": firstName,
      "last_name": lastName,
      "username": username,
      "email": email,
      "password": password,
      "accepted_terms_and_conditions": acceptedTerms,
    });
    return RegisterResponseModel.fromJson(response.data);
  }

  // POST /auth/login
  Future<UserModel> loginUser({
    required String email,
    required String password,
  }) async {
    final response = await api.post(ApiEndpoints.login, data: {
      "email": email,
      "password": password,
    });

    final data = response.data["data"] ?? {};
    final user = UserModel.fromJson(data);

    await SecureStorageService.saveUser(user);

    return user;
  }

  Future<void> logoutUser() async {
    await api.post('/auth/logout');

    await SecureStorageService.clearUser();
  }

  Future<void> verifyOtp(int userId, String otp) async {
    await api.post('/auth/verify_otp', data: {
      "user_id": userId,
      "otp": otp,
    });
  }

  // New method for account verification with OTP
  Future<VerifyAccountResponseModel> verifyAccount(
      int userId, String otp) async {
    final response =
        await api.post('${ApiEndpoints.verifyAccount}?isMobile=true', data: {
      "user_id": userId,
      "otp": otp,
    });
    return VerifyAccountResponseModel.fromJson(response.data);
  }

  // New method for resending OTP
  Future<VerifyAccountResponseModel> resendOtp(int userId) async {
    final response = await api.post(ApiEndpoints.resendOtp, data: {
      "user_id": userId,
    });
    return VerifyAccountResponseModel.fromJson(response.data);
  }

  Future<void> requestPasswordReset(String email) async {
    await api.post(
      '/auth/reset_password_request',
      queryParams: {"isMobile": true},
      data: {"email": email},
    );
  }

  Future<String> verifyResetPasswordOtp({
    required String email,
    required String otp,
  }) async {
    final response = await api.post('/auth/verify_reset_password_otp', data: {
      "email": email,
      "otp": otp,
    });
    if (response.data["status"] == "success") {
      return response.data["token"];
    } else {
      throw Exception(response.data["message"] ?? "OTP verification failed");
    }
  }

  Future<void> resetPassword({
    required String token,
    required String password,
    required String confirmPassword,
  }) async {
    await api.post(
      '/auth/reset_password',
      queryParams: {"token": token},
      data: {
        "password": password,
        "confirm_password": confirmPassword,
      },
    );
  }

  Future<void> changePassword({
    required int userId,
    required String oldPassword,
    required String newPassword,
    required String confirmNewPassword,
  }) async {
    await api.post('/auth/change_password', data: {
      "user_id": userId,
      "old_password": oldPassword,
      "new_password": newPassword,
      "confirm_new_password": confirmNewPassword,
    });
  }

  // Updated method name and parameter to match the API documentation.
  Future<void> resendVerificationEmail(String email) async {
    await api.post('/auth/resend-email-verification', data: {"email": email});
  }

  Future<Map<String, dynamic>> getCurrentUser() async {
    final response = await api.get('/auth/is_authenticated');

    if (response.data["status"] == "success" &&
        response.data["data"]?["authenticated"] == true) {
      return response.data; // Returning raw JSON instead of a UserModel
    } else {
      throw Exception("Not authenticated or error fetching user.");
    }
  }
}

File: frontend/app/services/auth-service.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { apiClient } from "../api/api-client";
import { ChangePasswordReqRequest } from "../types/auth";

export interface AuthResponse {
  status: string;
  message: string;
  data?: {
    authenticated?: boolean;
    user_id?: number;
    username?: string;
  };
}

export interface User {
  id: number;
  username: string;
}

export interface RegisterRequest {
  first_name: string;
  last_name: string;
  username: string;
  emailOrPhoneNumber: string;
  password: string;
  accepted_terms_and_conditions: boolean;
  home_location: string;
}

export interface RegisterResponse {
  status: string;
  message: string;
  data: {
    email: string;
    user_id: number;
    username: string;
  };
}

export interface VerifyOTPRequest {
  token: string;
}

export interface VerifyOTPResponse {
  status: string;
  message: string;
}

export interface LoginResponse {
  status: string;
  message: string;
  data: {
    user_id: number;
    token: string;
    email: string;
    last_name: string;
    first_name: string;
    username: string;
    profile_picture_url: string;
  };
}

export interface ResetPasswordReqRequest {
  email: string;
}

export interface ResetPasswordResponse {
  status: string;
  message: string;
}

export interface ResendEmailVerificationRequest {
  email: string;
}

export interface ResendEmailVerificationResponse {
  status: string;
  message: string;
}

export const authService = {
  /**
   * Check if the user is authenticated
   */
  async checkAuthentication(): Promise<{
    isAuthenticated: boolean;
    user: User | null;
  }> {
    try {
      const response = await apiClient.get<AuthResponse>(
        "/auth/is_authenticated"
      );
      if (response.status === "success" && response.data?.authenticated) {
        return {
          isAuthenticated: true,
          user: {
            id: response.data.user_id!,
            username: response.data.username!,
          },
        };
      }
      return { isAuthenticated: false, user: null };
    } catch (error) {
      console.error("Authentication check failed:", error);
      return { isAuthenticated: false, user: null };
    }
  },

  /**
   * Register a new user
   */
  async register(data: RegisterRequest): Promise<RegisterResponse> {
    try {
      const response = await apiClient.post<RegisterResponse>(
        "/auth/register",
        data
      );
      return response;
    } catch (error: any) {
      
      
      // Extract the actual error message from the API response
      const errorMessage = error?.response?.data?.message || 
                          error?.message || 
                          "Registration failed";
      
      throw new Error(errorMessage);
    }
  },

  /**
   * Resend OTP for verification
   */
  async resendOTP(
    user_id: string
  ): Promise<{ status: string; message: string }> {
    try {
      const response = await apiClient.post<{
        status: string;
        message: string;
      }>("/auth/resend_otp", { user_id });
      return response;
    } catch (error: any) {
      console.error("Resend OTP failed:", error);
      throw new Error(error?.message || "Resend OTP failed");
    }
  },

  async verifyEmail(data: VerifyOTPRequest): Promise<VerifyOTPResponse> {
    try {
      const response = await apiClient.post<VerifyOTPResponse>(
        "/auth/verify-account",
        data
      );
      return response;
    } catch (error: any) {
      console.error("Email verification failed:", error.message);
      throw new Error(
        error?.message || "Account verification failed"
      );
    }
  },

  async resendEmailVerification(
    data: ResendEmailVerificationRequest
  ): Promise<ResendEmailVerificationResponse> {
    try {
      const response = await apiClient.post<ResendEmailVerificationResponse>(
        `/auth/resend-email-verification`,
        { email: data.email }
      );
      return response;
    } catch (error: any) {
      console.error("Resend Email Verification failed:", error);
      throw new Error(
        error?.message || "Resend Email Verification failed"
      );
    }
  },
  
  async login(email: string, password: string): Promise<LoginResponse> {
    try {
      const response = await apiClient.post<LoginResponse>("/auth/login", {
        email,
        password,
      });
      if (response?.data?.user_id) {
        const userData = {
          user_id: response.data.user_id,
          username: response.data.username,
          email: response.data.email,
          last_name: response.data.last_name,
          first_name: response.data.first_name,
          profile_picture_url: response.data.profile_picture_url,
        };
        sessionStorage.setItem("user_id", response.data.user_id.toString());
        sessionStorage.setItem("user", JSON.stringify(userData));
      }
      return response;
    } catch (error: any) {
      console.error("Login error details:", error);
      // Extract only the user-friendly message, not the full API error
      const errorMessage = error?.response?.data?.message || 
                          error?.message || 
                          "Login failed. Please check your credentials and try again.";
      throw new Error(errorMessage);
    }
  },

  /**
   * Log out the current user
   */
  async logout(): Promise<boolean> {
    try {
      await apiClient.post("/auth/logout");
      return true;
    } catch (error) {
      console.error("Logout failed:", error);
      return false;
    }
  },

  /**
   * Reset password request
   */
  async resetPasswordRequest(
    data: ResetPasswordReqRequest
  ): Promise<ResetPasswordResponse> {
    try {
      const response = await apiClient.post<ResetPasswordResponse>(
        "/auth/reset_password_request",
        data
      );
      return response;
    } catch (error: any) {
      console.error("Reset Password failed:", error);
      throw new Error(
        error?.response?.data?.message || "Reset Password failed"
      );
    }
  },

  async changePassword(
    data: ChangePasswordReqRequest
  ): Promise<ResetPasswordResponse> {
    try {
      const response = await apiClient.post<ResetPasswordResponse>(
        `/auth/reset_password?token=${data.token}`,
        {
          password: data.password,
          confirm_password: data.confirm_password,
        }
      );
      return response;
    } catch (error: any) {
      console.error("Reset Password failed:", error);
      throw new Error(
        error?.response?.data?.message || "Reset Password failed"
      );
    }
  },

  /**
   * Fetch feed content (should this be in auth?)
   */
  async fetchFeed(): Promise<any> {
    try {
      const response = await apiClient.get("/content");
      return response;
    } catch (error: any) {
      console.error("Fetching feed failed:", error);
      throw new Error(error?.response?.data?.message || "Fetching feed failed");
    }
  },
};
